CMakeLists.txt

```diff
@@ -1,101 +1,151 @@
     set(CMAKE_BUILD_TYPE "Release" CACHE STRING "The type of build." FORCE)
 endif()
-if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang$" OR CMAKE_COMPILER_IS_GNUCXX)
-    if ("${CMAKE_CXX_FLAGS}" MATCHES "--coverage")
-        add_definitions(-DNDEBUG)
-    else()
-        include(CheckCXXCompilerFlag)
-        check_cxx_compiler_flag("-fsanitize=undefined" HAVE_FLAG_SANITIZE_UNDEFINED)
-        if(HAVE_FLAG_SANITIZE_UNDEFINED)
-            set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -fsanitize=undefined")
-        endif()
-    endif()
-    add_compile_options("-pipe" "-Wall" "-Wextra" "-Wconversion" "-Wpedantic")
-    if (EARCUT_WARNING_IS_ERROR)
-        add_compile_options("-Werror")
-    endif()
-    # optional: -march=native (builds with the optimizations available on the build machine (only for local use!))
-elseif(MSVC)
-    if (EARCUT_WARNING_IS_ERROR)
-        add_compile_options("/WX")
-    endif()
-endif()
 
-# earcut.hpp
-include_directories("include")
+
+include(GNUInstallDirs)
+
+add_library(earcut_hpp INTERFACE)
+add_library(earcut_hpp::earcut_hpp ALIAS earcut_hpp)
+
+target_include_directories(earcut_hpp INTERFACE
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+)
+
+set(CMAKE_CXX_STANDARD 11)
+
+if("${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION}" GREATER 3.7)
+    # Allow C++11 requirements to propagate when using recent CMake versions
+    target_compile_features(earcut_hpp INTERFACE cxx_std_11)
+endif()
 
 file(GLOB FIXTURE_SOURCE_FILES test/fixtures/*.cpp test/fixtures/*.hpp)
 source_group(fixtures FILES ${FIXTURE_SOURCE_FILES})
+add_library(fixtures OBJECT ${FIXTURE_SOURCE_FILES})
+target_compile_options(fixtures PRIVATE $<$<CXX_COMPILER_ID:MSVC>:/Od>)
+
+# In CMake 3.12, use target_link_libraries(fixtures PUBLIC earcut_hpp libtess2).
+# Since we support down to CMake 3.2, we need to manually propagate usage requirements of earcut_hpp
+target_include_directories(fixtures PRIVATE "$<TARGET_PROPERTY:earcut_hpp,INTERFACE_INCLUDE_DIRECTORIES>")
+target_compile_features(fixtures PRIVATE "$<TARGET_PROPERTY:earcut_hpp,INTERFACE_COMPILE_FEATURES>")
 
-if (MSVC)
-    set_source_files_properties(${FIXTURE_SOURCE_FILES} PROPERTIES COMPILE_FLAGS "/Od")
-endif()
 
 file(GLOB COMPARISON_SOURCE_FILES test/comparison/*.cpp test/comparison/*.hpp)
 source_group(comparison FILES ${COMPARISON_SOURCE_FILES})
+# this is interface since there is no cpp files in the comparison directory
+add_library(comparison INTERFACE)
+
 
 file(GLOB LIBTESS2_SOURCE_FILES test/comparison/libtess2/*.c test/comparison/libtess2/*.h)
-source_group(comparison\\libtess2 FILES ${LIBTESS2_SOURCE_FILES})
+source_group(comparison/libtess2 FILES ${LIBTESS2_SOURCE_FILES})
+add_library(libtess2 ${LIBTESS2_SOURCE_FILES})
+target_compile_options(libtess2 PRIVATE
+    $<$<CXX_COMPILER_ID:MSVC>:/wd4244 /wd4267>
+    $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-w>
```

Comment:
I'm unsure how these "NOT:$<CXX_COMPILER_ID:MSVC" conditions will hold up for more obscure compilers such as the Intel C+ compiler or IBM XLC++. But as these environments are not actively used or tested, I'll accept it for now.

Reply:
I did this to reflect the if that was before. It added `-w` on the else branch of `if(MSVC)`. The behavior should be the same as before, even if it was not optimal.